Source/RapyutaSimulationPlugins/Private/Tools/SimulationState.cpp

```diff
@@ -21,212 +22,241 @@
     {
-        if (IsValid(InEntities[InEntityName]))
+        // Check if Actor in EntitiesWithTag
+
+        bool ToEmplace = true;
+        for (AActor* TaggedActor : EntitiesWithTag[InTag].Actors)
         {
-            result = true;
+            if (TaggedActor->GetName() == Entity->GetName())
+            {
+                ToEmplace = false;
+            }
         }
-        else
+        if (ToEmplace)
         {
-            UE_LOG(LogRapyutaCore, Warning, TEXT("Request name %s entity gets invalid -> removed from Entities"), *InEntityName);
-            InEntities.Remove(InEntityName);
+            EntitiesWithTag[InTag].Actors.Emplace(Entity);
         }
     }
-    else if (bAllowEmpty && InEntityName.IsEmpty())
-    {
-        result = true;
-    }
     else
     {
-        UE_LOG(LogRapyutaCore,
-               Warning,
-               TEXT("%s is not under SimulationState control. Please call dedicated method to make Actors under "
-                    "SimulationState control."),
-               *InEntityName);
+        FRREntities actors;
+        actors.Actors.Emplace(Entity);
+        EntitiesWithTag.Emplace(InTag, actors);
     }
-
-    return result;
 }
 
-// https://isocpp.org/wiki/faq/templates#templates-defn-vs-decl
-template bool ASimulationState::CheckEntity<AActor*>(TMap<FString, AActor*>& InEntities,
-                                                     const FString& InEntityName,
-                                                     const bool bAllowEmpty);
-
-template bool ASimulationState::CheckEntity<TSubclassOf<AActor>>(TMap<FString, TSubclassOf<AActor>>& InEntities,
-                                                                 const FString& InEntityName,
-                                                                 const bool bAllowEmpty);
-
-bool ASimulationState::CheckEntity(const FString& InEntityName, const bool bAllowEmpty)
+void ASimulationState::AddSpawnableEntityTypes(TMap<FString, TSubclassOf<AActor>> InSpawnableEntityTypes)
 {
-    return CheckEntity<AActor*>(Entities, InEntityName, bAllowEmpty);
+    for (auto& elem : InSpawnableEntityTypes)
+    {
+        SpawnableEntityInfoList.Emplace(FRREntityInfo(elem));
+        SpawnableEntityTypes.Emplace(MoveTemp(elem.Key), MoveTemp(elem.Value));
+    }
 }
 
-bool ASimulationState::CheckSpawnableEntity(const FString& InEntityName, const bool bAllowEmpty)
+void ASimulationState::GetSpawnableEntityInfoList()
 {
-    return CheckEntity<TSubclassOf<AActor>>(SpawnableEntities, InEntityName, bAllowEmpty);
+    for (auto& elem : SpawnableEntityTypes)
+    {
+        SpawnableEntityInfoList.Emplace(FRREntityInfo(elem));
+    }
+    if (SpawnableEntityInfoList.Num() > 0)
+    {
+        GetWorld()->GetTimerManager().ClearTimer(TimerHandle);
+    }
 }
 
-void ASimulationState::GetEntityStateSrv(UROS2GenericSrv* InService)
+bool ASimulationState::ServerCheckSetEntityStateRequest(const FROSSetEntityState_Request& InRequest)

```

Comment:
why this check is necessary?

Reply:
To ensure that we don't perform the exact same request multiple times, this is more applicable to running this setup locally
as each robot's rosnode will receive the rosservice call and we will get 1 copy of the request for each robot